### PR TITLE
Add `no-for-loop` rule

### LINKED ISSUE #49: Rule proposal: `no-for-loop`
There's no reason to use normal `for` loops anymore for the common case.

The following:

``` js
const arr = ['foo', 'bar'];

for (let i = 0; i < arr.length; i++) {
  const el = arr[i];
  console.log(i, el);
}
```

Can now be written as:

``` js
const arr = ['foo', 'bar'];

for (const [i, el] of arr.entries()) {
  console.log(i, el);
}
```

Off-by-one errors is one of the most common bugs in software. [Swift actually removed this completely from the language.](https://github.com/apple/swift-evolution/blob/master/proposals/0004-remove-pre-post-inc-decrement.md)

We could probably ignore reverse loops and other cases, but the common case like above is better written in the latter form.
## 

@kevva @SamVerschueren @jfmengels @jamestalmage Thoughts?

### PR DESCRIPTION
**Pull request by futpib**
fixes #49

### CODE REVIEW

**Comment by sindresorhus** [comment]
Hey, thanks for the pull request 🙌 I have gotten a huge response on the IssueHunt bounty issues and I now have a very long backlog of pull requests to review, so it might take some time until I will be able to review this. However, you can help move this along by helping out and reviewing the [other submissions](https://github.com/sindresorhus/eslint-plugin-unicorn/pulls). If everyone helps everyone we can manage this 👌.

Some things to look for when reviewing:

- Ensure the rule addition adheres to the [instructions](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/new-rule.md).
- There are never too many tests. Try to find both passing and failing cases, especially obscure cases.
- Check the docs for typos and whether something could be more clear.
- Look at the rule code and see if anything could be simplified or done better. If something is unclear in the code, ask the submitter to explain.

**Comment by futpib** [comment]
@medusalix, @janowsiany, could you please help review this PR?

**Comment by medusalix** [comment]
@futpib I think the following case should be handled as well:

```js
for (let index = 0; index < array.length; index++);
```

There's currently an error being thrown because `body` is not defined for an `EmptyStatement`.

**Comment by futpib** [comment]
@manovotny :ok_hand: nice catch

File: docs/rules/no-for-loop.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Do not use for loop that can be replaced with a for-of loop
+
+There's no reason to use old school for loops anymore for the common case. You can instead use for-of loop (with `.entries()` if you need to access the index).
+
```
Comment (sindresorhus) [inline docs/rules/no-for-loop.md]: I think the docs here should mention something about this preventing off-by-one errors.

From the issue:

> Off-by-one errors is one of the most common bugs in software. [Swift actually removed this completely from the language.](https://github.com/apple/swift-evolution/blob/master/proposals/0004-remove-pre-post-inc-decrement.md)

**Comment by sindresorhus** [comment]
This is looking very good. I have been running it on many different repos of mine and it seems to be working great.

File: rules/no-for-loop.js [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+			const shouldFix = !someIdentifiersLeakOutOfTheLoop(node, [indexIdentifierName, elementIdentifierName], forScope, bodyScope);
+
+			if (shouldFix) {
+				problem.fix = fixer => {
+					const shouldGenerateIndex = isIndexIdentifierUsedElsewhereInTheLoopBody(node, indexIdentifierName, bodyScope);
+
+					const i = indexIdentifierName;
+					const el = elementIdentifierName;
+					const arr = arrayIdentifierName;
```
[the remark this answers is not in the document]
Comment (futpib) [reply rules/no-for-loop.js]: I introduced short variables just to make the `replacement` formatting below more readable:

```js
					const replacement = shouldGenerateIndex ?
						`const [${i}, ${el}] of ${arr}.entries()` :
						`const ${el} of ${arr}`;
```

vs

```js
					const replacement = shouldGenerateIndex ?
						`const [${indexIdentifierName}, ${elementIdentifierName}] of ${arrayIdentifierName}.entries()` :
						`const ${elementIdentifierName} of ${arrayIdentifierName}`;
```

Do you think the latter would be better?

[the remark this answers is not in the document]
Comment (sindresorhus) [reply rules/no-for-loop.js]: Would have been enough with `index`, `element`, `array`.

**Comment by futpib** [comment]
Is this still missing something?

**Comment by sindresorhus** [comment]
This is good to merge. Great work on this, @futpib 👌

**Pull request merged**